### ISSUE TITLE
[Bug]: CUDA out of memory 

### ISSUE THREAD

**Issue by marconovino** [hlky/stable-diffusion-webui#673]
### What happened?

When I try to generate 20 images at 1024x1024 with 50 sampling steps I get the following error:

`CUDA out of memory. Tried to allocate 16.00 GiB (GPU 0; 24.00 GiB total capacity; 10.18 GiB already allocated; 10.70 GiB free; 10.73 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF`

My rig:

I9-9900K
64GB DDR4 RAM
RTX 3090

### Version

0.0.1 (Default)

### What browsers are you seeing the problem on?

Chrome

### Where are you running the webui?

Windows

### Relevant log output

```shell
CUDA out of memory. Tried to allocate 16.00 GiB (GPU 0; 24.00 GiB total capacity; 10.18 GiB already allocated; 10.70 GiB free; 10.73 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by HanJammer** [comment]
Have you tried using --optimized or (even better) --optimized-turbo switch in the relauncher.py?

**Comment by Alphyn-gunner** [comment]
This is not a bug. 
If you read the error message carefully, you'll see that you're just about 2 gb short. Which is impressive. I bet if you tried a slightly lower resolution, like 960, it might actually even work.

**Comment by Nullav** [comment]
> Just an FYI: I don’t think 1024x1024 is achievable right now on consumer hardware. Even with a 3090.

It sorta is, actually. I say sorta, but the side effects are that gen gets dramatically slower even at 512x512, and the results at higher resolutions are predictably far more prone to duplication of elements.

**Comment by sedetweiler** [comment]
You are better off working at lower res and upsizing as you will get a ton of repeated elements at that resolution.

**Comment by ryudrigo** [comment]
If it's img2img, the repeated elements problem is usually not an issue (depends on the strength of the tranformation). Though, of course, just applying super resolution is bound to be a lot faster. I have not tested enough to know for sure the difference in quality of the two approaches

**Comment by mchaker** [comment]
Have you tried Doggettx's memory optimization?
https://gist.github.com/mchaker/e0cf180db6849fa03696293b3fc8f485

**Comment by ryudrigo** [comment]
Yes, by just looking at the code I get the impression that my optimization (https://github.com/basujindal/stable-diffusion/pull/122) is a little better. However, I didn't really test both on the same machine.

**Comment by mchaker** [comment]
I can test both. What would a test setup (parameters, etc) look like -- how would you like me to test it?

**Comment by ryudrigo** [comment]
Whoa thanks a lot!
please try generating one 1024x1024 image with 50 ddim steps an unet_bs (for the basujindal repository) of 2. n_samples (which is the batch size) should be 1. Please report on memory usage and time.
If you would be so kind, could you repeat the test for 512x512 images, changing n_samples (which is the batch size) to 6 and unet_bs to 12.
Thanks again =)

**Comment by mchaker** [comment]
Sure - can you provide exact commands for me to run? 😅  Thanks

**Comment by ryudrigo** [comment]
Took me some time to get the exact command, here you go:

python optimizedSD/optimized_txt2img.py --prompt "A painting of an apple" --H 1024 --W 1024 --seed 1280 --n_samples 2 --ddim_steps 50 --unet_bs 4

python scripts/txt2img.py --prompt "A painting of an apple" --H 1024 --W 1024 --seed 1280  --n_samples 2 --ddim_steps 50

These should generate two 1024x images each. It would be also helpful to know which GPU you are using

**Comment by mchaker** [comment]
Thanks for the commands! 

> It would be also helpful to know which GPU you are using

I will run the tests on both a Tesla P40 (24GB VRAM) and a Tesla P100 (16GB VRAM) and report back when I have results.

**Comment by ryudrigo** [comment]
I ended up downloading it myself out of expectation =P. Of course it's another card, so it would be interesting to know your results, too. Mine is an RTX 3060

basujindal (my PR) no turbo mode:
`python optimizedSD/optimized_txt2img.py --prompt "A painting of an apple" --H 1024 --W 1024 --seed 1280 --n_samples 2 --ddim_steps 50 --unet_bs 4 --sampler ddim`
2min 24 sec
5717MB

basujindal (my PR) with turbo mode:
`python optimizedSD/optimized_txt2img.py --prompt "A painting of an apple" --H 1024 --W 1024 --seed 1280 --n_samples 2 --ddim_steps 50 --unet_bs 4 --sampler ddim --turbo`
1 min  48 sec
6699MB

Doggettx:
`python scripts/txt2img.py --prompt "A painting of an apple" --H 1024 --W 1024 --seed 1280 --n_samples 2 --ddim_steps 50 --sampler ddim`
1min 49 sec
7993MB

So, overall, basujindal is better. However, the repositories are different in other places as well. A really thorough test would be changing just ldm/modules/attention.py and comparing. But those results have been enough for me so far.

**Comment by mchaker** [comment]
@ryudrigo I've finally completed the requested benchmarks:

Prompt: "A painting of an apple"

basujindal command: `export N_SAMPLES="2"; export UNET_BS="4"; export SD_IMG_SIZE="1024"; time python optimizedSD/optimized_txt2img.py --prompt "A painting of an apple" --H $SD_IMG_SIZE --W $SD_IMG_SIZE --seed 1280 --n_samples $N_SAMPLES --ddim_steps 50 --unet_bs $UNET_BS`

Doggettx command: `export N_SAMPLES="2"; export SD_IMG_SIZE="1024"; time python scripts/txt2img.py --prompt "A painting of an apple" --H $SD_IMG_SIZE --W $SD_IMG_SIZE --seed 1280 --n_samples $N_SAMPLES --ddim_steps 50`

stats:

optimization | card | image size | `n_samples` | `unet_bs` | memory (`nvidia-smi`) | `memory_final` | time
--- | --- | --- | --- | --- | --- | --- | ---
basujindal | Tesla P40 | 1024x1024 | 1 | 2 | 4773 MiB | 7.294464 | 231.6 sec
basujindal | Tesla P100 | 1024x1024 | 1 | 2 | 4889 MiB | 7.294464 | 197.4 sec
basujindal | Tesla P40 | 512x512 | 6 | 12 | 4099 MiB | 4.941312 | 162 sec
basujindal | Tesla P100 | 512x512 | 6 | 12 | 4215 MiB | 4.941312 | 156.6 sec
Doggettx | Tesla P40 | 1024x1024 | 1 | n/a | 24181 MiB | n/a | 454 sec
Doggettx | Tesla P100 | 1024x1024 | 1 | n/a | 13337 MiB | n/a | 293 sec
Doggettx | Tesla P40 | 512x512 | 6 | n/a | 17867 MiB | n/a | 268 sec
Doggettx | Tesla P100 | 512x512 | 6 | n/a | 11849 MiB | n/a | 200 sec

What I'm confused about, is how your optimization is better 😅  -- Doggettx's optimization seems to be more "thorough", but we have hard results now 🤔 

When I ran the Doggettx optimizations, the script seemed to run twice.... maybe the time should be cut in half?

![image](https://user-images.githubusercontent.com/10101544/189197898-8120ce94-6b4a-45f3-89c7-011692db4db9.png)

**Comment by ryudrigo** [comment]
After a lot of debugging and testing, I came to the conclusion that my way of optimizing is roughly equivalent (if implemented correctly) and slightly more complex than that of @Doggettx. In the end, I split the batch dimension and they split the last dimension (which is used for the softmax). I tried joining both approaches, but it didn't get any better.

So you can remove mine from the list, I'll mention this issue in the other repos I've made pull request to.

**Comment by mchaker** [comment]
Are you sure you want it removed? It's still a valid optimization

**Comment by ryudrigo** [comment]
in that case, I'll upload the corrections I've made and leave it as an option
